drivers/serial/uart_sf32lb.c

```diff
@@ -351,6 +396,363 @@ static void uart_sf32lb_irq_rx_disable(const struct device *dev)
 }
 #endif /* CONFIG_UART_INTERRUPT_DRIVEN */
 
+#ifdef CONFIG_UART_ASYNC_API
+static inline void uart_sf32lb_async_timer_start(struct k_work_delayable *work, size_t timeout)
+{
+	if ((timeout != SYS_FOREVER_US) && (timeout != 0)) {
+		LOG_DBG("Async timer started for %d us", timeout);
+		k_work_reschedule(work, K_USEC(timeout));
+	}
+}
+
+static void uart_sf32lb_dma_tx_done(const struct device *dma_dev, void *user_data, uint32_t channel,
+				    int status)
+{
+	struct uart_sf32lb_data *data = user_data;
+	const struct device *uart_dev = data->async.uart_dev;
+	const struct uart_sf32lb_config *config = uart_dev->config;
+	struct uart_event evt = {0};
+	unsigned int key = irq_lock();
+
+	k_work_cancel_delayable(&data->async.tx.timeout_work);
+
+	/* Disable DMA for TX */
+	sys_clear_bit(config->base + UART_CR3, USART_CR3_DMAT_Pos);
+
+	evt.type = UART_TX_DONE;
+	evt.data.tx.buf = data->async.tx.buf;
+	evt.data.tx.len = data->async.tx.len;
+	if (data->async.cb) {
+		data->async.cb(uart_dev, &evt, data->async.user_data);
+	}
+
+	/* Reset TX Buffer */
+	data->async.tx.buf = NULL;
+	data->async.tx.len = 0U;
+
+	irq_unlock(key);
+}
+
+static void uart_sf32lb_dma_rx_done(const struct device *dma_dev, void *user_data, uint32_t channel,
+				    int status)
+{
+	struct uart_sf32lb_data *data = user_data;
+	const struct device *uart_dev = data->async.uart_dev;
+	const struct uart_sf32lb_config *config = uart_dev->config;
+
+	struct uart_event evt = {0};
+	unsigned int key = irq_lock();
+	int err = 0;
+
+	/* Disable DMA for RX */
+	sys_clear_bit(config->base + UART_CR3, USART_CR3_DMAR_Pos);
+
+	/* Notify RX_RDY */
+	evt.type = UART_RX_RDY;
+	evt.data.rx.buf = data->async.rx.buf;
+	evt.data.rx.len = data->async.rx.len;
+	evt.data.rx.offset = 0;
+	if (data->async.cb && evt.data.rx.len) {
+		data->async.cb(uart_dev, &evt, data->async.user_data);
+	}
+
+	/* Release current buffer */
+	evt.type = UART_RX_BUF_RELEASED;
+	evt.data.rx_buf.buf = data->async.rx.buf;
+	if (data->async.cb) {
+		data->async.cb(uart_dev, &evt, data->async.user_data);
+	}
+
+	/* Load next buffer and request another */
+	data->async.rx.buf = data->async.rx.next_buf;
+	data->async.rx.len = data->async.rx.next_len;
+	data->async.rx.next_buf = NULL;
+	data->async.rx.next_len = 0U;
+
+	/* If there's a next buffer, restart DMA */
+	if (data->async.rx.buf) {
+		sf32lb_dma_reload_dt(&config->rx_dma, (uintptr_t)(config->base + UART_RDR),
+				     (uint32_t)data->async.rx.buf, data->async.rx.len);
+
+		err = sf32lb_dma_start_dt(&config->rx_dma);
+		if (err) {
+			LOG_ERR("Error starting Rx DMA (%d)", err);
```

Comment:
silent error, maybe return? Also, depending on scenario/dma driver, you may just ignore DMA API return codes if you know it's not gonna fail `(void)sf32lb_dma...`